### PR TITLE
Retry loading ResourceManager resources in all cases

### PR DESCRIPTION
**Pull request by yihezkel**
### Changed
Retry loading ResourceManager resources in all cases

### CODE REVIEW

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java [hunk 1]
```diff
@@ -96,74 +101,237 @@ public void close() {
...
-                        timer.schedule(new RefreshIngestionResourcesTask(), refreshTimeOnFailure);
+                    return null;
+                }, "ResourceManager.refreshIngestionResource");
+            } catch (Exception e) {
+                log.error("Error in refreshIngestionResources. " + e.getMessage(), e);
+                scheduleRefreshIngestionResourcesTask(refreshTimeOnFailure);
+            }
+        }
+
+        private void refreshIngestionResources() throws IngestionClientException, IngestionServiceException {
```
Comment (yihezkel) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java:124]: All these methods were on the `ResourceManager` class, though now `RefreshIngestionResourcesTask` is effectively their wrapper with retry logic, so they belong there.

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java [hunk 2]
```diff
@@ -96,74 +101,237 @@ public void close() {
...
+            // Here we use tryLock(): If there is another instance doing the refresh, then just skip it.
+            if (ingestionResourcesLock.writeLock().tryLock()) {
+                try {
+                    log.info("Refreshing Ingestion Resources");
+                    IngestionResourceSet newIngestionResourceSet = new IngestionResourceSet();
+                    Retry retry = Retry.of("get ingestion resources", retryConfig);
```
Comment (yihezkel) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java:130]: Note we already have some retry logic

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java [hunk 3]
```diff
@@ -96,74 +101,237 @@ public void close() {
...
+                        while (table.next()) {
+                            String resourceTypeName = table.getString(0);
+                            String storageUrl = table.getString(1);
+                            addIngestionResource(newIngestionResourceSet, resourceTypeName, storageUrl);
+                        }
                     }
+                    populateStorageAccounts(newIngestionResourceSet);
+                    ingestionResourceSet = newIngestionResourceSet;
+                    refreshedAtLeastOnce = true;
+                    ingestionResourcesRefreshedOnce.signalAll();
```
Comment (yihezkel) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java:146]: Notify any threads awaiting the first/next refresh.

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java [hunk 4]
```diff
@@ -43,20 +43,23 @@
...
+import java.util.concurrent.locks.Condition;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.stream.Stream;
 
 class ResourceManager implements Closeable, IngestionResourceManager {
-    public static final String SERVICE_TYPE_COLUMN_NAME = "ServiceType";
     private static final long REFRESH_INGESTION_RESOURCES_PERIOD = TimeUnit.HOURS.toMillis(1);
-    private static final long REFRESH_INGESTION_RESOURCES_PERIOD_ON_FAILURE = TimeUnit.MINUTES.toMillis(15);
+    private static final long REFRESH_INGESTION_RESOURCES_PERIOD_ON_FAILURE = TimeUnit.MINUTES.toMillis(1);
```
Comment (yihezkel) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java:53]: Do we want to wait 15 minutes between retries if there was an error?

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java [hunk 5]
```diff
@@ -96,74 +101,237 @@ public void close() {
...
 
-        timer.schedule(new RefreshIngestionAuthTokenTask(), 0);
-        timer.schedule(new RefreshIngestionResourcesTask(), 0);
+    private void init() {
+        scheduleRefreshIngestionResourcesTask(0L);
+        scheduleRefreshIngestionAuthTokenTask(0L);
+    }
+
+    // If we combined these 2 methods, ensuring we're using distinct synchronized locks for both would be inelegant
+    private synchronized void scheduleRefreshIngestionResourcesTask(Long delay) {
```
Comment (yihezkel) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java:264]: Making the method `synchronized` ensures the task is never `null` and is always scheduled.

File: ingest/src/test/java/com/microsoft/azure/kusto/ingest/ResourceManagerTimerTest.java [hunk 6]
```diff
@@ -41,16 +42,58 @@ class BooleanHolder {
         });
 
         ResourceManager resourceManager = new ResourceManager(mockedClient, 1000L, 500L, null);
+        assertFalse(resourceManager.refreshIngestionResourcesTask.isRefreshedAtLeastOnce());
+        assertFalse(resourceManager.refreshIngestionAuthTokenTask.isRefreshedAtLeastOnce());
+
         int runtime = 0;
         while (!booleanHolder.gotHere && runtime < 5000) {
             Thread.sleep(100);
```
Comment (yihezkel) [inline ingest/src/test/java/com/microsoft/azure/kusto/ingest/ResourceManagerTimerTest.java:50]: Not the biggest fan of calling `Thread.sleep()`. TODO for a future enhancement...

**Review by AsafMah — APPROVED** [review]
Looks good, well documented.
Just please update CHANGELOG.md

**Comment by yihezkel** [comment]
@ohadbitt 
I made a lot of changes here, and you've worked in this area before, so just a head's up in case you want to review.

@AsafMah
I made a few changes since your approval, so a head's up in case you want to rereview.

**Review by AsafMah — APPROVED** [review]

**Pull request merged**